### PR TITLE
WIP: POC for removing useless curves in revision graph

### LINKED ISSUE #5782: Readability revision graph
**Do you want to request a *feature* or report a *bug*?**
Discussion/feature

**What is the current behavior?**
The revision graph is very curly. I think the readability could be improved easily. I propose to make the change as visible in this screenprint (left is new):
image

POC: #5783

**Environment you encounter the issue:**
- Git Extensions 3.00.rc2
- d6496b811db20f8890cb3a50caa14937f3e0b650  (Dirty)
- Git 2.19.1.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3110.0
- DPI X:100,00% Y:100,00%

### PR DESCRIPTION
**Pull request by spdr870**
Fixes #5782 

Changes proposed in this pull request:
- Example implementation for reducing useless curves at nearly 0 performance cost

First commit, remove the not-needed curves:
image 

Second commit, make the lines more smooth:
image

What did I do to test the code and ensure quality:
- Scrolling through the log

Has been tested on (remove any that don't apply):
- GIT 2.19
- Windows 10

### CODE REVIEW

File: GitUI/UserControls/RevisionGrid/Columns/RevisionGraphColumnProvider.cs [hunk 1]
```diff
@@ -370,42 +375,45 @@ private Brush GetBrushForRevision(RevisionGraphRevision revisionGraphRevision, b
...
-                    var yMid = (y0 + y1) / 2f;
-                    var c0 = new PointF(offset + p0.X, offset + yMid);
-                    var c1 = new PointF(offset + p1.X, offset + yMid);
-                    var e0 = new PointF(offset + p0.X, offset + p0.Y);
-                    var e1 = new PointF(offset + p1.X, offset + p1.Y);
-                    g.DrawBezier(lanePen, e0, c0, c1, e1);
-                }
+                // Anti-aliasing with bezier & PixelOffsetMode.HighQuality
+                // introduces an offset of ~1/8 px - compensate it.
+                g.SmoothingMode = SmoothingMode.AntiAlias;
```
Comment (drewnoakes) [inline GitUI/UserControls/RevisionGrid/Columns/RevisionGraphColumnProvider.cs]: The comment still exists, but I don't see the 1/8px offset in code.

File: GitUI/UserControls/RevisionGrid/Columns/RevisionGraphColumnProvider.cs [hunk 2]
```diff
@@ -370,42 +375,45 @@ private Brush GetBrushForRevision(RevisionGraphRevision revisionGraphRevision, b
...
+            float uncurlyFactor = 7.5f;
+            var p0 = new PointF(x0, y0); // Start
+            var p1 = new PointF((x0 + (x1 * uncurlyFactor) + x2) / (uncurlyFactor + 2f), (y0 + (y1 * uncurlyFactor) + y2) / (uncurlyFactor + 2f)); // Center
+            var p2 = new PointF(x2, y2); // End
+
+            using (var lanePen = new Pen(laneBrush, LaneLineWidth))
+            {
+                // Anti-aliasing with bezier & PixelOffsetMode.HighQuality
+                // introduces an offset of ~1/8 px - compensate it.
+                g.SmoothingMode = SmoothingMode.AntiAlias;
```
Comment (drewnoakes) [inline GitUI/UserControls/RevisionGrid/Columns/RevisionGraphColumnProvider.cs]: Ditto.

**Comment by mstv** [comment]
Thank you, the graph looks _much_ clearer now!

Two little things could be improved yet.
1. The curves should always go exactly vertically through nodes, i.e. the ends of diagonal sections should be rounded so that they converge towards the vertical line, like at the green arrow in the screenshot. This will make outgoing and incoming curves symmetrical.
grafik
2. Maybe these one-row-bows can be avoided, too:
grafik

**Comment by gerhardol** [comment]
@spdr870 
What is remaining before this can be merged?

**Comment by GintasS** [comment]
This is an old PR and should be closed or set to a draft. @gerhardol

**Comment by gerhardol** [comment]
Several of the changes in this PR has been included in PRs by @mstv. Can you summarize if something is not included?
